File: Week03/46.全排列.py
#
# @lc app=leetcode.cn id=46 lang=python3
#
# [46] 全排列
#

# @lc code=start
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        # time: 36ms, beats 94.71%
        res = [[]]
        for num in nums:
            tmp_res = []
            for sub_set in res:
                for i in range(len(sub_set)+1):
                    tmp_res.append(sub_set[:i] + [num] + sub_set[i:])    # insert [num]  append!
            res = tmp_res
            # 想用列表生成式，没成功。。。
        return res
# @lc code=end

File: Week03/46.全排列.java
/*
 * @lc app=leetcode.cn id=46 lang=java
 *
 * [46] 全排列
 */

// @lc code=start
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        // time: 2ms beats 81.74%
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if (nums.length == 0 ) return res;
        List<Integer> list0 = new ArrayList<Integer>();
        list0.add(nums[0]);
        res.add(list0);

        for (int i = 1; i < nums.length; ++i) {
            List<List<Integer>> new_res = new ArrayList<List<Integer>>();
            for (int j = 0; j <= i; ++j) {
                for (List<Integer> list : res) {
                    List<Integer> new_list = new ArrayList<Integer>(list);
                    new_list.add(j, nums[i]);
                    new_res.add(new_list);
                }
            }
            res = new_res;
        }
        return res;
    }
}
// @lc code=end

